### DISCUSSION TITLE
How to get form-data parser to save files with original extension

### DISCUSSION THREAD

**Discussion by mdkq** [oakserver/oak#472] [Q&A] [answered]
I have this code:
```js
  .post("/message", async (ctx) => {
    const body = await ctx.request.body({ type: "form-data" }).value.read({
      outPath: `${Deno.cwd()}/static/private/attachment`,
    });
    console.log(body);
  });
```
which works great for most files. However sometimes the browser doesn't know what mime type to set for a given extension (i.e. custom extensions, or even some known extensions like .ttf/.otf), and when I send a file with no type to oak via `FormData`, it seems oak defaults to inferring content type `application/octet-stream` and saving the file with a `.bin` extension. I don't care about the content type since I am just saving the files, but how can I get oak to preserve the original extension?

edit: loving oak btw thanks to everyone for the great work

**Comment by kitsonk** [comment, marked as answer]
For security reasons, whenever the file is saved locally, it uses the provided content type to determine the extension.  It is a lot better to trust browsers than end users when providing files.  Trusting the uploaded file name and its extension is a problem, as they are easy to manipulate and may not be valid for the target operating/filesystem.

Therefore the `FormDataFile` provides both the `.filename` and `.originalName` as part of the structure (as well as the provided `.contentType`), so if you "trust" what is uploaded, you can do post processing to reconcile it, using whatever logic you are comfortable with from a security and usability perspective.

Adding an option to automatically do this doesn't make sense to me personally, as it is really important that the user of oak considers the implications of post processing uploaded files, in a way that makes the most sense for their use case.

**Reply by mdkq** [reply]
Very interesting thank you
